src/payment/coupon/data/sagas.test.js

```diff
@@ -0,0 +1,246 @@
+import { runSaga } from 'redux-saga';
+
+import { configureApiService } from './service';
+import saga, { handleAddCoupon, handleRemoveCoupon } from './sagas';
+import { addCoupon, addCouponBegin, addCouponSuccess, addCouponFailure, removeCouponBegin, removeCouponSuccess, removeCoupon, removeCouponFailure } from './actions';
+import { PERCENTAGE_BENEFIT } from './constants';
+
+describe('saga tests', () => {
+  describe('main saga', () => {
+    it('should setup its sub sagas correctly', () => {
+      const gen = saga();
+
+      expect(gen.next()).toMatchSnapshot();
+      expect(gen.next()).toMatchSnapshot();
+      expect(gen.next()).toEqual({ done: true, value: undefined });
+    });
+  });
+
+  describe('handleAddCoupon', () => {
```

Comment:
But for the action handling sagas... game on!  This test mocks out `apiClient` for the service layer and then runs the actual saga against an action, proving that it dispatches its downstream actions or exceptions properly.  The saga calls the service methods which do their whole thing and get mock data back from the mocked apiClient.  It's a lot of setup, but it really exercises the sagas and services, which is nice.